components/viaduct/src/backend/ffi.rs

```diff
@@ -0,0 +1,190 @@
+    // it. Besides, we already own it.
+    let response_bytes = response.into_vec();
+
+    let response: msg_types::Response = match Message::decode(&response_bytes) {
+        Ok(v) => v,
+        Err(e) => {
+            panic!(
+                "Failed to parse protobuf returned from fetch callback! {}",
+                e
+            );
+        }
+    };
+
+    if let Some(exn) = response.exception {
+        let exn_name = exn
+            .name
+            .as_ref()
+            .map(|s| s.as_str())
+            .unwrap_or("<unknown exception>");
+        let exn_msg = exn
+            .msg
+            .as_ref()
+            .map(|s| s.as_str())
+            .unwrap_or("<no message provided>");
+        log::error!(
+            // Well, we caught *something* java wanted to tell us about, anyway.
+            "Caught network error (presumably). Type: {}, message: {:?}",
+            exn_name,
+            exn_msg
+        );
+        return Err(Error::NetworkError(format!("{}: {:?}", exn_name, exn_msg)));
+    }
+    let status = response
+        .status
+        .ok_or_else(|| backend_error!("Missing HTTP status"))?;
+
+    if status < 0 || status > i32::from(u16::max_value()) {
+        return Err(backend_error!("Illegal HTTP status: {}", status));
+    }
+
+    let mut headers = crate::Headers::with_capacity(response.headers.len());
+    for (name, val) in response.headers {
+        let hname = match crate::HeaderName::new(name) {
+            Ok(name) => name,
+            Err(e) => {
+                // Ignore headers with invalid names, since nobody can look for them anyway.
+                log::warn!("Server sent back invalid header name: '{}'", e);
+                continue;
+            }
+        };
+        // Not using Header::new intentionally, since the error it returns is
+        // for request headers.
+        headers.insert_header(crate::Header {
+            name: hname,
+            value: val,
+        });
+    }
+
+    let url = url::Url::parse(
+        &response
+            .url
+            .ok_or_else(|| backend_error!("Response has no URL"))?,
+    )
+    .map_err(|e| backend_error!("Response has illegal URL: {}", e))?;
+
+    Ok(crate::Response {
+        url,
+        request_method: method,
+        body: response.body.unwrap_or_default(),
+        status: status as u16,
+        headers,
+    })
+}
+
+/// Type of the callback we need callers on the other side of the FFI to
+/// provide.
+///
+/// Takes and returns a ffi_support::ByteBuffer. (TODO: it would be nice if we could
+/// make this take/return pointers, so that we could use JNA direct mapping. Maybe
+/// we need some kind of ThinBuffer?)
+///
+/// This is a bit weird, since it requires us to allow code on the other side of
+/// the FFI to allocate a ByteBuffer from us, but it works.
+///
+/// The code on the other side of the FFI is responsible for freeing the ByteBuffer
+/// it's passed using `viaduct_destroy_bytebuffer`.
```

Comment:
Out of curiosity, why is it the caller's responsibility to free the input ByteBuffer?  Since the call is synchronous, it feels like we could let the caller "borrow" the bytebuffer and then free it ourselves when it returns.  To be clear, I ask because I expect to learn something interesting, not because I think this is doing it wrong :-)

Reply:
Primarily to keep things consistent with how bytebuffer is used elsewhere (and how the documentation it has says it should be used), but this also has lower peak memory usage -- instead of having a buffer for the request and the buffer around for the response at the same time, we only have the one that is active around.